### PR TITLE
Add a helper method to dereference arrays

### PR DESCRIPTION
**Pull request by ausi**
Implements #5100

> We should then check the whole codebase where we manually use `unset()` to remove references and use this new method there.

I didn’t find where we had problems with references in the past, looks like we did not use the `unset()` fix there 😕
So maybe we should move this to 5.0 and use it only in https://github.com/contao/contao/blob/f79bca7f82f5afd339ea70de6492724955e07b7c/core-bundle/src/DataContainer/DataContainerOperation.php#L34-L36

### CODE REVIEW

**Review by m-vo — DISMISSED** [review]
The changes look good to me. I think I'd prefer to only merge it into 5.0 for now and backport it if we find any more occurrences in code originating <5.0. But I'm fine with merging it into 4.9 as well. :relaxed: .

**Comment by ausi** [comment]
> So maybe we should move this to 5.0 and use it only in …

Done

**Review by m-vo — DISMISSED** [review]
LGTM :+1: 

I wonder if we should handle the self-reference case?
```php
$array = [&$array];

StringUtil::resolveReferences($array); // results in an endless loop
```

**Comment by ausi** [comment]
> I wonder if we should handle the self-reference case?

And throw an exception in that case?
How can we detect them?

File: core-bundle/contao/library/Contao/StringUtil.php [hunk 1]
```diff
@@ -1129,4 +1129,12 @@ public static function numberToString(float|int $number, int|null $precision = n
...
 		return rtrim("$signPart$wholePart.$decimalPart", '.');
 	}
+
+	public static function resolveReferences(array $array): array
+	{
+		return array_map(
+			static fn ($value) => \is_array($value) ? self::resolveReferences($value) : $value,
+			$array,
+		);
+	}
```
Comment (m-vo) [inline core-bundle/contao/library/Contao/StringUtil.php:1139]: Something like this should probably work to detect self references:

```suggestion
    public static function resolveReferences(array $array): array
    {
        return self::doResolveReferences($array);
    }

    private static function doResolveReferences(array $array, array $traversed = []): array {

        return array_map(
            static function ($value) use ($array, $traversed) {
                if (in_array($array, $traversed, true)) {
                    throw new \InvalidArgumentException('Cannot dereference an array that contains a reference to itself.');
                }

                $traversed[] = $array;

                return \is_array($value) ? self::doResolveReferences($value, $traversed) : $value;
            },
            $array,
        );
    }
```

But I'm not sure if we should consider this edge case or keep it simple.

Comment (ausi) [reply core-bundle/contao/library/Contao/StringUtil.php:1139]: The `in_array()` check would be pretty expensive I think. I’d probably be in favor of keeping it simple and fast ☺️

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**